Speedup CI

NobodyXu opened the pull request:
## Motivation

tokio CI is quite slow.

## Solution

 - Use `cargo-nextest` to speedup testing
 - Use `cargo-check` instead of `cargo-build` and `cargo test --no-run`.
 - Run `cargo-clippy` with `--no-deps`
 - Set `CARGO_UNSTABLE_SPARSE_REGISTRY` to `true` so that nightly toolchain would use sparse crates.io registries
 - Adjust order of steps to avoid caching crates installed by `taiki-e/install-action@v2`
 - Use `Swatinem/rust-cache@v2` in job no-atomic-u64
 - Add concurrenty group to cancel outdated CI


Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

taiki-e commented on .github/workflows/ci.yml:
```diff
@@ -311,6 +346,8 @@ jobs:
         env:
           RUSTFLAGS: --cfg tokio_unstable -Dwarnings
 
+  # TODO: Once https://github.com/cross-rs/cross/issues/716 lands,
+  # switch to using cargo-nextest with cross to speedup tests
```
FWIW, I guess my [setup-cross-toolchain-action](https://github.com/taiki-e/setup-cross-toolchain-action) (currently used in no-atomic-u64 job) should already work with nextest.

NobodyXu replied:
I've also changed job `cross-check` to use setup-cross-toolchain-action.

taiki-e reviewed (changes requested):
Thanks!

taiki-e commented:
(I canceled old CI runs on this PR)

NobodyXu commented on .github/workflows/ci.yml:
```diff
@@ -88,24 +89,33 @@ jobs:
+          set -euxo pipefail
+          cargo nextest run --features full
+          cargo test --doc --features full
+        shell: bash
```
Thanks, I've removed this.

taiki-e commented on .github/workflows/ci.yml:
```diff
@@ -331,21 +367,36 @@ jobs:
+      # Install linker and libraries for i686-unknown-linux-gnu
+      - uses: taiki-e/setup-cross-toolchain-action@v1
+        with:
+          target: i686-unknown-linux-gnu
```
```suggestion
          target: ${{ matrix.target }}
```

NobodyXu replied:
Thanks, I've fixed this.

taiki-e commented on .github/workflows/ci.yml:
```diff
@@ -331,21 +367,36 @@ jobs:
+        uses: taiki-e/install-action@v2
+        with:
+          tool: cargo-nextest
+      # Install linker and libraries for i686-unknown-linux-gnu
```
```suggestion
      # Install linker and libraries
```

NobodyXu replied:
Ehhh I've removed the comment.

Feel free to open another review if you want to add it back.

taiki-e commented on .github/workflows/ci.yml:
```diff
@@ -301,16 +335,22 @@ jobs:
-        uses: taiki-e/install-action@cross
-      - run: cross check --workspace --all-features --target ${{ matrix.target }}
+
+      - uses: taiki-e/setup-cross-toolchain-action@v1
+        with:
+          target: ${{ matrix.target }}
```
Since tokio and tokio-* are libraries, cross-check probably does not require the installation of the toolchain.

NobodyXu replied:
I reversed the change because `setup-cross-toolchain-action` does not support target `arm-linux-androideabi`.

NobodyXu replied:
> Since tokio and tokio-* are libraries, cross-check probably does not require the installation of the toolchain.

Hmmm, that will work only if tokio does not depend on any C/C++ libraries.

NobodyXu replied:
Let me have a try.

NobodyXu replied:
Thanks, it worked.

taiki-e commented:
Also, the arm-unknown-linux-gnueabihf in cross-test probably needs to be changed to armv5te-unknown-linux-gnueabi. It seems it was for testing platforms without AtomicU64, which is incorrect because arm-unknown-linux-gnueabihf has AtomicU64. ARM targets that do not have AtomicU64 are armv5te-unknown-linux-gnueabi and arm-linux-**androideabi**.

https://github.com/tokio-rs/tokio/blob/70364b707975b84daf223915d60a1610de2539a2/.github/workflows/ci.yml#L323-L329

NobodyXu commented:
> Also, the arm-unknown-linux-gnueabihf[1](#user-content-fn-1-6978474c97b3907bd9217dba8d7d891d) in cross-test probably needs to be changed to armv5te-unknown-linux-gnueabi. It seems it was for testing platforms without AtomicU64, which is probably incorrect because arm-unknown-linux-gnueabihf has AtomicU64. ARM targets that do not have AtomicU64 are armv5te-unknown-linux-{gnu,musl,uclibc}eabi and arm-linux-**androideabi** (and some bere-metal targets).

Thank you, I've updated PR as you requested.

NobodyXu commented on .github/workflows/ci.yml:
```diff
@@ -117,9 +129,9 @@ jobs:
-      # Build benchmarks. Run of benchmarks is done by bench.yml workflow.
-      - name: build benches
-        run: cargo build --benches
+      # Check benchmarks. Run of benchmarks is done by bench.yml workflow.
```
@taiki-e job `test` is taking quite long to run on windows.

Shall I split this into a separate job?

taiki-e replied:
Splitting into separate jobs seems reasonable to me.

NobodyXu replied:
Done!

NobodyXu commented on .github/workflows/ci.yml:
```diff
@@ -320,32 +354,46 @@ jobs:
         env:
           RUSTFLAGS: --cfg tokio_unstable -Dwarnings --cfg tokio_no_ipv6 ${{ matrix.rustflags }}
       # Now run without parking_lot
       - name: Remove `parking_lot` from `full` feature
         run: sed -i '0,/parking_lot/{/parking_lot/d;}' tokio/Cargo.toml
       # The `tokio_no_parking_lot` cfg is here to ensure the `sed` above does not silently break.
-      - run: cross test -p tokio --features full,test-util --target ${{ matrix.target }} --tests
+      - run: |
+          set -euxo pipefail
+          cargo nextest run -p tokio --features full,test-util --target ${{ matrix.target }}
+          cargo test --doc -p tokio --features full,test-util --target ${{ matrix.target }}
         env:
           RUSTFLAGS: --cfg tokio_unstable -Dwarnings --cfg tokio_no_ipv6 --cfg tokio_no_parking_lot ${{ matrix.rustflags }}
```
@taiki-e job cross-test is also taking quite long.
Shall I split this part into a separate job?

NobodyXu replied:
Done

taiki-e commented on .github/workflows/ci.yml:
```diff
@@ -320,32 +354,47 @@ jobs:
+      - name: First tests run with all features (including parking_lot)
+        run: |
+          set -euxo pipefail
+          cargo nextest run -p tokio --all-features --target ${{ matrix.target }}
+          cargo test --doc -p tokio --all-features --target ${{ matrix.target }}
```
Hmm. nextest+setup-cross-toolchain-action ([log](https://github.com/tokio-rs/tokio/actions/runs/4981204722/jobs/8915183782)) seems to be slower than cross ([log](https://github.com/tokio-rs/tokio/actions/runs/4980067925/jobs/8912721614)).

I guess it's one of the following two things:

- The parallelization performed by libtest is causing a problem with QEMU, and setting `RUST_TEST_THREADS=1` environment variable fixes the problem.
- The parallelization performed by nextest is causing a problem with QEMU, and using `cargo test` instead of `cargo nextest run` fixes the problem.

(In the current CI run, setup-cross-toolchain-action is a bit [slow](https://github.com/tokio-rs/tokio/actions/runs/4981204722/jobs/8915183928) because it depends on the state of Azure's APT mirror used in GitHub Actions runners, but it is usually [much faster](https://github.com/taiki-e/portable-atomic/actions/runs/4969849868/jobs/8893446429).)

NobodyXu replied:
> The parallelization performed by nextest is causing a problem with QEMU, and using `cargo test` instead of `cargo nextest` run fixes the problem.

AFAIK `cargo-nextest` runs each test in a separate process, so this should not be a problem.

It's likely `cargo test --doc` is a problem, passing `--test-threads=1` to it might fix the problem.

NobodyXu replied:
I've added `-- --test-threads=1` to all invocation of `cargo-test` in `no-atomics` and `cross-test*`.

NobodyXu replied:
I've also added in to job `asan` where I tried using `cargo-nextest` again.

taiki-e commented:
BTW, the performance of Valgrind-related tests is currently unstable and [sometimes takes a long time](https://github.com/tokio-rs/tokio/actions/runs/4981133581/jobs/8914961821), but I guess this should be fixed by passing `--fair-sched=yes` to Valgrind.

https://github.com/tokio-rs/tokio/blob/70364b707975b84daf223915d60a1610de2539a2/.github/workflows/stress-test.yml#L41
https://github.com/tokio-rs/tokio/blob/70364b707975b84daf223915d60a1610de2539a2/.github/workflows/ci.yml#L173
https://github.com/tokio-rs/tokio/blob/70364b707975b84daf223915d60a1610de2539a2/.github/workflows/ci.yml#L182

taiki-e commented on .github/workflows/ci.yml:
```diff
@@ -9,6 +9,7 @@ name: CI
 env:
   RUSTFLAGS: -Dwarnings
   RUST_BACKTRACE: 1
+  CARGO_UNSTABLE_SPARSE_REGISTRY: true
```
Thanks for the explanation. Could you leave a TODO comment because I would like to remove that environment variable when the nightly toolchain is updated?

NobodyXu replied:
DONE

NobodyXu commented:
> BTW, the performance of Valgrind-related tests is currently unstable and [sometimes takes a long time](https://github.com/tokio-rs/tokio/actions/runs/4981133581/jobs/8914961821), but I guess this should be fixed by passing `--fair-sched=yes` to Valgrind.

Thanks, I've added this as instructed.

NobodyXu commented:
According to my observation of recent CI runs, this PR reduces workflow `ci.yml` from ~30m => ~18m

NobodyXu commented:
@taiki-e I've rebased this PR and removed the unstable sparse registries environment variable

NobodyXu commented:
cc @taiki-e @carllerche This PR has been stale for quite some time.
Can you review this please?

Thanks!

taiki-e commented on .github/workflows/ci.yml:
```diff
@@ -327,118 +410,199 @@ jobs:
+      - name: Tests run with all features (including parking_lot)
+        run: |
+          set -euxo pipefail
+          cargo nextest run -p tokio --all-features --target ${{ matrix.target }}
+          cargo test --doc -p tokio --all-features --target ${{ matrix.target }} -- --test-threads 1
```
Assuming that limiting libtest's parallelism is the correct solution for https://github.com/tokio-rs/tokio/pull/5691#discussion_r1193940213, we probably need to limit libtest's parallelism for both, not only for doctest (IIUC nextest runs multiple test binaries in separate processes, but [does nothing about libtest's parallelism themselves](https://github.com/taiki-e/cargo-llvm-cov/issues/258#issuecomment-1516625520).). Setting the RUST_TEST_THREADS=1 environment variable is a reasonable way to do that.

Well, if that is the correct solution, we should set the environment variable on the setup-cross-toolchain-action side.

NobodyXu replied:
Thanks!

I've applied your suggestion and updated the PR!

taiki-e commented on .github/workflows/stress-test.yml:
```diff
@@ -28,14 +28,14 @@ jobs:
       # Runs each of the examples using Valgrind. Detects leaks and displays them.
       - name: Run valgrind
-        run: valgrind --error-exitcode=1 --leak-check=full --show-leak-kinds=all ./target/release/examples/${{ matrix.stress-test }}
+        run: valgrind --error-exitcode=1 --leak-check=full --show-leak-kinds=all ./target/release/examples/${{ matrix.stress-test }} --fair-sched=yes
```
```suggestion
        run: valgrind --error-exitcode=1 --leak-check=full --show-leak-kinds=all --fair-sched=yes ./target/release/examples/${{ matrix.stress-test }}
```

taiki-e commented on .github/workflows/ci.yml:
```diff
@@ -647,24 +832,18 @@ jobs:
+          targets: wasm32-wasi
 
       # Install dependencies
       - name: Install cargo-hack
```
```suggestion
      - name: Install cargo-hack, wasmtime, and cargo-wasi
```

taiki-e commented on .github/workflows/ci.yml:
```diff
@@ -327,118 +410,199 @@ jobs:
         env:
           RUSTFLAGS: --cfg tokio_unstable --cfg tokio_taskdump -Dwarnings --cfg tokio_no_atomic_u64
+
+  no-atomic-u64-check:
```
Why was this separated? I thought i686 was not so slow because it runs without QEMU.

NobodyXu replied:
It's because
 - `no-atomic-u64` takes quite some time
 - Running `cargo-check` in `no-atomic-u64` disrupts the caching, since they use different `RUSTFLAGS`.

NobodyXu commented on .github/workflows/ci.yml:
```diff
@@ -156,22 +219,22 @@ jobs:
       # Run with valgrind
       - name: Run valgrind test-mem
-        run: valgrind --error-exitcode=1 --leak-check=full --show-leak-kinds=all ./target/debug/test-mem
+        run: valgrind --error-exitcode=1 --leak-check=full --show-leak-kinds=all ./target/debug/test-mem --fair-sched=yes
```
Agreed, will adopt the suggestion.

NobodyXu commented:
@Darksonn Seems like disabling feature `mio/log` by default breaks [CI `wasm32-wasi`](https://github.com/tokio-rs/tokio/actions/runs/5112829534/jobs/9191341712)

Darksonn commented:
Yes, see https://github.com/tokio-rs/mio/pull/1676.

NobodyXu commented:
Thanks @Darksonn !
I would rebase this PR once the fix has landed in tokio

NobodyXu commented:
@carllerche Newly introduced self-tuning and `unstable::test_disable_lifo_slot` seems to be quite slow when running on qemu, often stuck for a really long time.

taiki-e commented:
It seems test on i686 with Mutex-based AtomicU64 is hanging (it does not use QEMU): https://github.com/tokio-rs/tokio/actions/runs/5222355294/jobs/9427778431?pr=5691

NobodyXu commented:
> It seems test on i686 with Mutex-based AtomicU64 is hanging (it does not use QEMU): https://github.com/tokio-rs/tokio/actions/runs/5222355294/jobs/9427778431?pr=5691

So we don't need that `taiki-e/setup-cross-toolchain-action@v1` in `no-atomic-test`?

taiki-e commented:
> So we don't need that `taiki-e/setup-cross-toolchain-action@v1` in `no-atomic-test`?

No, it installs linker and libraries for i686 as the comment says.

https://github.com/tokio-rs/tokio/blob/c5d52c17ae800f1d39d7132fd9c452f2fae0e168/.github/workflows/ci.yml#L393-L394

NobodyXu commented:
> > So we don't need that `taiki-e/setup-cross-toolchain-action@v1` in `no-atomic-test`?
> 
> No, it installs linker and libraries for i686 as the comment says.
> 
> https://github.com/tokio-rs/tokio/blob/c5d52c17ae800f1d39d7132fd9c452f2fae0e168/.github/workflows/ci.yml#L393-L394

Thanks!
I've removed it from `no-atomic-check` though since it only runs `cargo-check` and also removed `RUST_TEST_THREADS=1` there since it's not run using `qemu`.

taiki-e commented:
> It seems test on i686 with Mutex-based AtomicU64 is hanging (it does not use QEMU): https://github.com/tokio-rs/tokio/actions/runs/5222355294/jobs/9427778431?pr=5691

We previously used `--test-threads=1` for this test, and I think it is possible that the removal of that may have caused the problem.

https://github.com/tokio-rs/tokio/blob/c5d52c17ae800f1d39d7132fd9c452f2fae0e168/.github/workflows/ci.yml#L397

NobodyXu commented:
> We previously used `--test-threads=1` for this test, so I think it is possible that the removal of that may have caused the problem.
> 
> https://github.com/tokio-rs/tokio/blob/c5d52c17ae800f1d39d7132fd9c452f2fae0e168/.github/workflows/ci.yml#L397

Ok, I've added it back.

taiki-e reviewed (approved):
LGTM, thanks @NobodyXu!

NobodyXu commented:
CI "Test hyper (macos-latest)" failed due to:

```
thread 'http2_keep_alive_count_server_pings' panicked at 'timed out waiting for pings: Elapsed(())', tests/server.rs:2488:10
```

looks like just a temporary network failure.

taiki-e commented:
Hmm, is nextest not handling custom targets correctly? In addition to the following warning, all dependencies are recompiled when doctest is built.

```
warning: could not determine target triple: error deserializing target triple from --target option
caused by: unknown platform triple
caused by: unknown triple string: target-specs/i686-unknown-linux-gnu.json
caused by: Unrecognized architecture: target
```

https://github.com/tokio-rs/tokio/actions/runs/5244982846/jobs/9472222567#step:7:1274

Although it probably wouldn't be difficult for nextest to add support for custom targets (probably basically just duplicating the logic around [this](https://github.com/taiki-e/cargo-config2/blob/c7a341139d68c32c25be827ddaa543f1512fbe94/src/resolve.rs#L487-L505)).

NobodyXu commented:
@taiki-e I opened a feature request in nextest https://github.com/nextest-rs/nextest/issues/887

NobodyXu commented:
@taiki-e cargo-nextest v0.9.54 has released which support custom target.

I triggered the PR before new release of install-action is cut which didn't include cargo-nextest v0.9.54

can you re-run the tests so that it uses latest nextest?

NobodyXu commented:
Merge conflict resolved, this PR is now mergeable.

taiki-e commented:
Thanks!
